### PR TITLE
Introduce ccache: a compiler cache

### PR DESCRIPTION
**Pull request by glevkovich**
* **Problem Overview**  

Introduce ccache: a compiler cache

https://ccache.dev/
Ccache is a compiler cache. It speeds up recompilation by caching
previous compilations and detecting when the same compilation is
being done again. Ccache is free software, released under the GNU
General Public License version 3 or later. See also the license page.

Important points to mention:
*) ccache existed on concord-bft docker image but wasn't used
till now, and the version was old. Upgraded to newest version 4.6.1.
*) For now, ccache is configured to work only locally using Makefile
dockerized build, or by using CMake.
*) ccache configuration file is under .ccache/ccache.conf
*) The cache is stored under $HOME/.ccache. It is positioned there
to cache information
from multiple workspaces/projects.
*) ccache is enabled by default. To disable ccache, run build with
CONCORD_BFT_CMAKE_CCACHE=FALSE or set disable in ccache.conf to false
*) There are 2  targets added/modified in Makefile:
create-ccache-folder: is used internally to create ccache folder
clean: will clean build folder by default, or you can clean everything
with the flag CLEAN_ALL set to TRUE. If you like to clean only untracked
files use CLEAN_UNTRACKED=TRUE. If you like to clean ccache cache use 
CLEAN_CACHE=TRUE.
*) For native linux, It is assumed that caller has called cmake and
install_deps.sh.

* **Testing Done**  
Mainly manual tests, to see cache really works.

### CODE REVIEW

File: .ccache/ccache.conf [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# For complete documentation, see https://ccache.dev/manual/4.6.1.html#_configuration
+
+# This option specifies the maximum size of the cache. Use 0 for no limit. The default value is 5G.
+# Available suffixes: k, M, G, T (decimal) and Ki, Mi, Gi, Ti (binary). The default suffix is G
+max_size = 10G
+
+# This option specifies where ccache will keep its cached compiler outputs.
+cache_dir = /mnt/ccache/
```
Comment (WildFireFlum) [inline .ccache/ccache.conf:8]: shouldn't this be in the build directory or under bft's root?

File: .ccache/ccache.conf [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+#If true, ccache will compress data it puts in the cache. However, this option has no effect on how files are
+#retrieved from the cache; compressed and uncompressed results will still be usable regardless of this option.
+# The default is true.
+compression = true
+
+# This option determines the level at which ccache will compress object files using the real-time compression
+# algorithm Zstandard. It only has an effect if compression is enabled (which it is by default). In restricted space environments,
+# it is recommended to set the value of this parameter to 19 (or via environment parameter CCACHE_COMPRESSLEVEL)
+compression_level = 1
```
Comment (WildFireFlum) [inline .ccache/ccache.conf:18]: why compress if you use the lowest possible compression rate?

File: .ccache/ccache.conf [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+# When true, ccache will just call the real compiler, bypassing the cache completely. The default is false.
+disable = false
+
+# If set to a file path, ccache will write information on what it is doing to the specified file.
+# This is useful for tracking down problems.
+log_file = /mnt/ccache/ccache.log
```
Comment (WildFireFlum) [inline .ccache/ccache.conf:25]: Should probably be local

File: .ccache/ccache.conf [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+# If set to a file path, ccache will write statistics counter updates to the specified file. This is useful
+# for getting statistics for individual builds. To show a summary of the current stats log, use ccache --show-log-stats.
+stats_log = /mnt/ccache/ccache_stats.log
```
Comment (WildFireFlum) [inline .ccache/ccache.conf:29]: should be local

File: .ccache/ccache.conf [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+# If true, ccache will not discard the comments before hashing preprocessor output.
+keep_comments_cpp = true
```
Comment (WildFireFlum) [inline .ccache/ccache.conf:32]: Shouldn't this be set to false?

File: Makefile [hunk 6]
```diff
@@ -157,7 +166,7 @@ login: ## Login to the container. Note: if the container is already running, log
 	fi
 
 .PHONY: gen-cmake
-gen-cmake: ## Generate cmake files, used internally
+gen-cmake: create-ccache-folder ## Generate cmake files, used internally
```
Comment (WildFireFlum) [inline Makefile:169]: If the cache directory is under bft, you can use mkdir to create it instead of a separate target

File: Makefile [hunk 7]
```diff
@@ -326,20 +335,28 @@ test-single-gtest-case: ## Run a single GoogleTest test case: `make test-single-
...
 .PHONY: clean
-clean: ## Clean Concord-BFT build directory
+clean: create-ccache-folder ## Clean Concord-BFT build directory
 	docker run ${BASIC_RUN_PARAMS} \
 		${CONCORD_BFT_CONTAINER_SHELL} -c \
 		"rm -rf ${CONCORD_BFT_BUILD_DIR}"
 
 .PHONY: clean-all
-clean-all: ## Clean Concord-BFT build directory and any other untracked/ignored files.For a 'dry run' use DRY_RUN=TRUE, for an interactive run use INTER=TRUE.
+clean-all: ## Clean Concord-BFT build directory and any other untracked/ignored files. Does not include ccache cache folder.For a 'dry run' use DRY_RUN=TRUE, for an interactive run use INTER=TRUE.
```
Comment (WildFireFlum) [inline Makefile:344]: I believe that this should also remove the ccache (same effect as removing the build directory), otherwise the name is misleading.

File: install_deps.sh [hunk 8]
```diff
@@ -41,7 +40,8 @@ apt-get update && apt-get ${APT_GET_FLAGS} install \
     sudo \
     vim \
     iproute2 \
-    wget
+    wget \
+    libhiredis-dev
```
Comment (WildFireFlum) [inline install_deps.sh:44]: Does ccache use redis?

File: install_deps.sh [hunk 9]
```diff
@@ -363,6 +363,23 @@ git clone -b v0.7.0 --depth 1 https://github.com/jaegertracing/jaeger-client-cpp
...
+cd "$HOME"
+wget https://github.com/ccache/ccache/releases/download/v${CCACHE_VER}/ccache-${CCACHE_VER}.tar.xz && \
+  tar xvf ccache-${CCACHE_VER}.tar.xz && \
+  cd ccache-${CCACHE_VER} && \
+  mkdir build && cd build && \
+  cmake -DCMAKE_BUILD_TYPE=Release .. && make && make install && \
+  cd "${HOME}" && \
+  rm -rf ccache-${CCACHE_VER} ccache-${CCACHE_VER}.tar.xz
+CCACHE_PATH=$(which ccache)
+ln -s "${CCACHE_PATH}" /usr/local/bin/gcc
```
Comment (WildFireFlum) [inline install_deps.sh:377]: do you have to create these links manually? Isn't there a ccache script that does it?

**Review by WildFireFlum — CHANGES_REQUESTED** [review]

**Review by eyalrund — APPROVED** [review]

File: Makefile [hunk 10]
```diff
@@ -72,6 +73,13 @@ ifeq (${CONCORD_BFT_CMAKE_CODECOVERAGE},TRUE)
 	CONCORD_BFT_CMAKE_CXX_FLAGS_RELEASE='-O0 -g'
 endif
 
+# The consistency parameter makes sense only at MacOS.
```
Comment (toly-kournik) [inline Makefile:76]: How is this related to ccache?

Comment (toly-kournik) [inline Makefile:76]: What about native build?

**Comment by toly-kournik** [comment]
Just want to ensure ccache works with a native build

File: .ccache/ccache.conf (hunk 1 above)
Comment (glevkovich) [reply .ccache/ccache.conf:8]: No, as his is done on purpose.
First, we need to remember that /mnt/ccache is a folder mount from the host (for containerized).
Also, the default ccache recommended cache is under ${HOME}, see manual.
All the idea of ccache is to boost your compilations time between projects and between builds after 'make clean.
If you put it in the build you are losing most of the idea of the cache. The cache will be deleted when you clean, or when you make a new project or workspace it won't help anything + it will be duplicated in different projects. And I want to share it between projects.

For example, I have 3 vmewathena where I work on vmwathenaX/concord/submodule/concord-bft/, and I also have 3 concord-bft separate workspaces/projects (total 6 projects). When I clean them, I don't want to clean the cache (which is what will happen when calling 'make clean') and I want to share a single cache among 6 projects + I don't want a 'make clean' to clean my cache. So, the idea is that the cache is not part of the build folder. To delete the cache you call make 'clean-ccache-cache'

File: .ccache/ccache.conf (hunk 2 above)
Comment (glevkovich) [reply .ccache/ccache.conf:18]: - Lowest compression rate still compress it more than 80% and the space on local build is not unlimited, so I wanted to have compression still.
- I first had it on 19 (higest for all versions) to check how it will be on CI. The differnce in speed in 5-10% 19 vs 1.
- I prefer speed over space for local, and space over speed on CI. still, 1 gives around 80% compression on very minimal cost 1-3% of the speed.

File: .ccache/ccache.conf (hunk 4 above)
Comment (glevkovich) [reply .ccache/ccache.conf:29]: I'm going to move both log files into concord-bft/build/ccache/

File: .ccache/ccache.conf (hunk 3 above)
Comment (glevkovich) [reply .ccache/ccache.conf:25]: I'm going to move both log files into concord-bft/build/ccache/

File: .ccache/ccache.conf (hunk 5 above)
Comment (glevkovich) [reply .ccache/ccache.conf:32]: The default is false, my idea was that ccache processor shouldn't touch the file content at all, just hash them. Anyway, I don't think it matters much.

File: Makefile (hunk 6 above)
Comment (glevkovich) [reply Makefile:169]: cach folder (as we discussed) should not be under bft

File: install_deps.sh (hunk 8 above)
Comment (glevkovich) [reply install_deps.sh:44]: This is for future use - secondary cache.Rethinking - I'm going to remove it for now. If we want to use secondary redis cache in the future, we just add the support.

File: install_deps.sh (hunk 9 above)
Comment (glevkovich) [reply install_deps.sh:377]: CMAKE_CXX_COMPILER_LAUNCHER/CMAKE_C_COMPILER_LAUNCHER might crate the links but then the image is removed and the linsk too, so if we want permanent links we must do this as part of the image.
Still, without ovverriding the path nothing will work, so it's ok to put the links here even if ccache is disabled.

**Comment by glevkovich** [comment]
It’s a problem to set as ${HOME}/.ccache inside container since we are not running as root when building. we are running as user id 1000 (non root) so we dont have permission to that folder but it is still set as out home folder.

File: .ccache/ccache.conf (hunk 3 above)
Comment (glevkovich) [reply .ccache/ccache.conf]: Hi, I've decided to remove this log completely for now for multiple reasons. In case we need to debug i'll add it.

File: .ccache/ccache.conf (hunk 4 above)
Comment (glevkovich) [reply .ccache/ccache.conf]: Hi, I've decided to remove this log completely for now for multiple reasons. In case we need to debug i'll add it.

File: Makefile (hunk 7 above)
Comment (glevkovich) [reply Makefile:345]: That's a bit of a problem. What if I just want to clean the whole local folder, build and any leftovers and not clean ccache? This is usually what I want. I agree name might be a bit misleading but the idea is to clean everything inside the current root project folder, and ccache cache is not under that folder, so it's a matter of interpertation - everything is cleaned under concord-bft, we don't delete /tmp/* as well, right?

ccache deletion should be explicit by running clean-ccache-cache or rm -rf $HOME/.ccache/*

File: install_deps.sh (hunk 8 above)
Comment (glevkovich) [reply install_deps.sh]: removed this installed.

File: .ccache/ccache.conf [hunk 11]
```diff
@@ -0,0 +1,26 @@
+# For complete documentation, see https://ccache.dev/manual/4.6.1.html#_configuration
+# Comments:
+# 1) By default, cache_dir is in $HOME/.ccache
```
Comment (cloudnoize) [inline .ccache/ccache.conf:3]: Should the cache_dir be inside the build dir?

File: Makefile [hunk 12]
```diff
@@ -326,20 +336,28 @@ test-single-gtest-case: ## Run a single GoogleTest test case: `make test-single-
...
 .PHONY: clean
-clean: ## Clean Concord-BFT build directory
+clean: create-ccache-folder ## Clean Concord-BFT build directory
 	docker run ${BASIC_RUN_PARAMS} \
 		${CONCORD_BFT_CONTAINER_SHELL} -c \
 		"rm -rf ${CONCORD_BFT_BUILD_DIR}"
 
 .PHONY: clean-all
-clean-all: ## Clean Concord-BFT build directory and any other untracked/ignored files.For a 'dry run' use DRY_RUN=TRUE, for an interactive run use INTER=TRUE.
+clean-all: ## Clean Concord-BFT build directory and any other untracked/ignored files. Does not include ccache cache folder.For a 'dry run' use DRY_RUN=TRUE, for an interactive run use INTER=TRUE.
```
Comment (cloudnoize) [inline Makefile:345]: Why it does not include the cache dir?
 When we clean, we want a fresh build without any use of cache and reuse of objects no?

File: .ccache/ccache.conf (hunk 11 above)
Comment (glevkovich) [reply .ccache/ccache.conf:3]: This is done intentionally.They must be inside this dir for few reasons,  (this is also the recommended default by ccache) with the main ones:

- To supshare cache between multiple projects/workspaces. Not everyone work with Git worktrees.
- Even when you clean or when you clone, ccache will go here. ccache is transperent to cleaning the build folder.

File: Makefile (hunk 12 above)
Comment (glevkovich) [reply Makefile:345]: Well, my idea says there are 3 levels of clean, but I guess there is no single truth here. What makes sense to me might not make sense to others:

- 'make clean' is regular clean of project bin -we dont want to delete ccache since we want to clean local build foder, and keep shared cache with other projects.
- make 'clean-all' - clean build and all repository artifacts. it still cleans things under CURRENT project. ccache is above prohject/folder/concord-bft level, it is a shared (or remote) cache shared by all users (projects, hosts or users)
- make 'clean-ccache-cache' - the only expclit command which will clean the ccache, and for me it makes sensse since cleaning the whole shared cache for everyone shouldn't be on the project level. this one goes to $HOME/.ccache/ by default and delete everything under.

Comment (glevkovich) [reply Makefile:345]: Im going to add explentation in confluence about it.
It think my logic make sense here. Other people might find a different solution more appropriate.

Comment (glevkovich) [reply Makefile]: I think all the many cleans created confusion.
Tere going to be a single make clean which cleans only build folder. Using flags it will be possible to clean different things. For more infor see confluene page or make help.

**Review by arc-vmware — APPROVED** [review]
LGTM!!!
Thanks @glevkovich for adding this. Will it be a good idea to add this in [Readme](https://github.com/vmware/concord-bft/blob/master/README.md)? If it's available in make help then its fine.

File: CMakeLists.txt [hunk 13]
```diff
@@ -29,6 +29,7 @@ option(TXN_SIGNING_ENABLED "Enable External concord client transcattion signing"
 option(BUILD_COMM_TCP_PLAIN "Enable TCP communication" OFF)
 option(LEAKCHECK "Enable Address and Leak Sanitizers" OFF)
 option(THREADCHECK "Enable Thread Sanitizer" OFF)
+option(USE_CCACHE "Enable ccache to speed up recompilation" ON)
```
Comment (toly-kournik) [inline CMakeLists.txt:32]: this doesn't work with a native build in a conjunction with the `if (USE_CCACHE)` later in this file.
I propose not to change `CMakeLists.txt` at all, but rather call it as recommended `cmake -DCMAKE_CXX_COMPILER_LAUNCHER=ccache ..` from an outside `Makefile`

File: install_deps.sh [hunk 14]
```diff
@@ -401,51 +400,70 @@ wget ${WGET_FLAGS} https://archive.apache.org/dist/thrift/0.11.0/thrift-0.11.0.t
...
+    tar xvf ccache-${CCACHE_VER}.tar.xz && \
+    cd ccache-${CCACHE_VER} && \
+    mkdir build && cd build && \
+    cmake -DCMAKE_BUILD_TYPE=Release -DREDIS_STORAGE_BACKEND=OFF .. && \
+    make && \
+    make install && \
+    cd "${HOME}" && \
+    rm -rf ccache-${CCACHE_VER} ccache-${CCACHE_VER}.tar.xz
+  CCACHE_PATH=$(which ccache)
+  ln -s "${CCACHE_PATH}" /usr/local/bin/gcc
```
Comment (toly-kournik) [inline install_deps.sh:460]: why make this symbolic links?
the compiler command should look like:
`ccache /usr/local/bin/clang++ ...` making it possible to choose between enabling/disabling ccache usage and choosing a compiler

Comment (glevkovich) [reply install_deps.sh:460]: There are 2 ways to run ccache and the symbolic link is a valid one. ccache knows how to find the compiler according to the nvoked sym link name, so there is no way it will not run the right g++/clang++.
To simplify native build, It's better to have it without sym links. I'll check if it's possible to run it the way you suggest..

Comment (glevkovich) [reply install_deps.sh:460]: We don't need to use ccache explcitly. I checked that. the CMake CMAKE_CXX_COMPILER_LAUNCHER definition do the trick, all we need to do is to pass it as CMake definition when building

**Review by toly-kournik — APPROVED** [review]
LGTM from the point of view of a native build!

**Review by WildFireFlum — APPROVED** [review]
